### PR TITLE
Column saving per character

### PR DESCRIPTION
**Pull request by Ner-f**
Added a feature that saves column selection to a cookie for each character.

## Description

Each character will save selected columns to a cookie using the character name.

## Related Issue

#113 

## Motivation and Context

Users will no longer need to change their column layout when they are switching between character types.

## How Has This Been Tested?

I created multiple character types with their own column data. Made sure it loads properly and removes properly when changing character names and deleting characters.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: www/src/public/js/controllers/builder/main.js [hunk 1]
```diff
@@ -507,6 +507,13 @@
             $scope.editCharacterModel = {"name": $scope.allLists[listIndex].name};
             var list = $scope.selectedList;
 
+            // load savedColumns from character cookie
+            var cookieConsent = $cookies.get("cookie-consent");
+            if (cookieConsent) {
+                var columnCookie = $cookies.get($scope.allLists[listIndex].name);
```
Comment (steventhorne) [inline www/src/public/js/controllers/builder/main.js]: Can we prepend this with `sc` (selected columns). So it'll end up looking like:
```js
var columnCookie = $cookies.get(`sc-${$scope.allLists[listIndex].name}`);
```

File: www/src/public/js/controllers/builder/main.js [hunk 2]
```diff
@@ -677,6 +678,13 @@
                 selectListByName();
             }
 
+            // load columns
+            if (cookieConsent) {
+                console.log($scope.allLists[$scope.selectedListIndex].name);
+                var columnCookie = $cookies.get($scope.allLists[$scope.selectedListIndex].name);
```
Comment (steventhorne) [inline www/src/public/js/controllers/builder/main.js]: prepend `sc-` here, too

File: www/src/public/js/controllers/builder/main.js [hunk 3]
```diff
@@ -1425,7 +1433,21 @@
         /** Confirm function for the edit character modal. */
         var editCharacter = function() {
             if (validateEditCharacter()) {
-                $scope.allLists[$scope.selectedListIndex].name = $scope.textInputModalModel.input;
+                var cookieConsent = $cookies.get("cookie-consent");
+                var oldName = $scope.allLists[$scope.selectedListIndex].name
```
Comment (steventhorne) [inline www/src/public/js/controllers/builder/main.js:1439]: prepend `sc-` here, too

File: www/src/public/js/controllers/builder/main.js (hunk 1 above)
Comment (steventhorne) [reply www/src/public/js/controllers/builder/main.js]: what happens if they don't have this cookie yet? for instance, when they first load the page after this update? what columns will it use? can we default to the old cookie? That cookie is still used for the item search page, so it should still exist

**Review by steventhorne — CHANGES_REQUESTED** [review]
Looks good overall, just a couple of minor nitpicks

File: www/src/public/js/controllers/builder/main.js (hunk 1 above)
Comment (Ner-f) [reply www/src/public/js/controllers/builder/main.js]: I added a quick test for this but it seems to be loading a default column list from somewhere else.

Comment (Ner-f) [reply www/src/public/js/controllers/builder/main.js]: I see what happens..

I moved where it loads the saved column string so it happens _after_ the selectListByName function gets called. I had to do this so I could get the selected list name. When the select list function runs it also calls the loadSelectedColumns function which will initially pass a null value for columnCookie and show all default columns.

File: www/src/public/js/controllers/builder/main.js [hunk 4]
```diff
@@ -507,6 +507,22 @@
...
+            var cookieConsent = $cookies.get("cookie-consent");
+            var cookieDate = new Date();
+            cookieDate.setFullYear(cookieDate.getFullYear() + 20);
+            if (cookieConsent) {
+                var columnCookie = $cookies.get(`sc-${$scope.allLists[$scope.selectedListIndex].name}`);
+                if(!columnCookie) {
+                    var columnCookie = $cookies.get("sc2");
+                    if(!columnCookie) { //if cookie sc2 does not exist, insert a basic column string
+                        $cookies.put("sc2", "Slot-Name-Align-Rent-Str-Min-Dex-Con-Per-Spi-Ac-", {path: "/", samesite: "lax", secure: true, expires: cookieDate});
+                        var columnCookie = "Slot-Name-Align-Rent-Str-Min-Dex-Con-Per-Spi-Ac-";
```
Comment (steventhorne) [inline www/src/public/js/controllers/builder/main.js:519]: You could move this one line up and reuse it in the `$cookies.put` call.

File: www/src/public/js/controllers/builder/main.js [hunk 5]
```diff
@@ -677,6 +687,12 @@
                 selectListByName();
             }
 
+            // load columns
```
Comment (steventhorne) [inline www/src/public/js/controllers/builder/main.js]: Is this needed? Since we're calling `selectListByName` right before this and that will load the cookie in the `selectListVariantByIndex` func. At least I'm pretty sure this is true. You should verify

File: www/src/public/js/controllers/builder/main.js [hunk 6]
```diff
@@ -727,6 +734,15 @@
                 $cookies.remove("scl1");
             }
 
+            //save columns
```
Comment (steventhorne) [inline www/src/public/js/controllers/builder/main.js:731]: Why was this moved from a few lines above?

File: www/src/public/js/controllers/builder/main.js [hunk 7]
```diff
@@ -1425,7 +1441,21 @@
...
+                var oldName = $scope.allLists[$scope.selectedListIndex].name
+                var newName = $scope.textInputModalModel.input;
+
+                //copy savedColumns to new cookie and remove the old cookie
+                if(cookieConsent) {
+                    var cookieDate = new Date();
+                    cookieDate.setFullYear(cookieDate.getFullYear() + 20);
+                    var savedColumns = $cookies.get(oldName);
+                    $cookies.put(`sc-${newName}`, savedColumns, {path: "/", samesite: "lax", secure: true, expires: cookieDate});
+                    $cookies.remove(`sc-${oldName}`, {"path": "/"});
```
Comment (steventhorne) [inline www/src/public/js/controllers/builder/main.js:1448]: If the name doesn't change, this will cause the cookie to be removed completely. Should probably only do this whole cookie update if `oldName !== newName`

**Review by steventhorne — CHANGES_REQUESTED** [review]
Almost there

File: www/src/public/js/controllers/builder/main.js (hunk 6 above)
Comment (Ner-f) [reply www/src/public/js/controllers/builder/main.js:731]: It was causing an undefined error for $scope.selectedListIndex. IIRC

File: www/src/public/js/controllers/builder/main.js (hunk 5 above)
Comment (Ner-f) [reply www/src/public/js/controllers/builder/main.js]: It is not needed it seems.

**Review by steventhorne — APPROVED** [review]
